src/control/cmd/daos/attribute.go

```diff
@@ -111,68 +118,117 @@ func listDaosAttributes(hdl C.daos_handle_t, at attrType, verbose bool) ([]*attr
 		return nil, err
 	}
 
-	var err error
+	if verbose {
+		return getDaosAttributes(hdl, at, attrNames)
+	}
+
 	attrs := make([]*attribute, len(attrNames))
 	for i, name := range attrNames {
-		if verbose {
-			attrs[i], err = getDaosAttribute(hdl, at, name)
-			if err != nil {
-				return nil, errors.Wrap(err, "list attributes failed")
-			}
-		} else {
-			attrs[i] = &attribute{Name: name}
-		}
+		attrs[i] = &attribute{Name: name}
 	}
 
 	return attrs, nil
+
 }
 
-func getDaosAttribute(hdl C.daos_handle_t, at attrType, name string) (*attribute, error) {
-	var rc C.int
-	var attrSize C.size_t
+// getDaosAttributes fetches the values for the given list of attribute names.
+// Uses the bulk attribute fetch API to minimize roundtrips.
+func getDaosAttributes(hdl C.daos_handle_t, at attrType, names []string) (attrList, error) {
+	if len(names) == 0 {
+		attrList, err := listDaosAttributes(hdl, at, false)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to list attributes")
+		}
+		names = make([]string, len(attrList))
+		for i, attr := range attrList {
+			names[i] = attr.Name
+		}
+	}
+	numAttr := len(names)
 
-	attrName := C.CString(name)
-	defer freeString(attrName)
+	// First, build a slice of C strings for the attribute names.
+	attrNames := make([]*C.char, numAttr)
+	for i, name := range names {
+		attrNames[i] = C.CString(name)
+	}
+	defer func(nameSlice []*C.char) {
+		for _, name := range nameSlice {
+			freeString(name)
+		}
+	}(attrNames)
 
+	// Next, create a slice of C.size_t entries to hold the sizes of the values.
+	// We have to do this first in order to know the buffer sizes to allocate
+	// before fetching the actual values.
+	attrSizes := make([]C.size_t, numAttr)
+	var rc C.int
 	switch at {
 	case poolAttr:
-		rc = C.daos_pool_get_attr(hdl, 1, &attrName, nil, &attrSize, nil)
+		rc = C.daos_pool_get_attr(hdl, C.int(numAttr), &attrNames[0], nil, &attrSizes[0], nil)
 	case contAttr:
-		rc = C.daos_cont_get_attr(hdl, 1, &attrName, nil, &attrSize, nil)
+		rc = C.daos_cont_get_attr(hdl, C.int(numAttr), &attrNames[0], nil, &attrSizes[0], nil)
 	default:
 		return nil, errors.Errorf("unknown attr type %d", at)
 	}
 	if err := daosError(rc); err != nil {
-		return nil, errors.Wrapf(err, "failed to get attribute %q", name)
-	}
-
-	attr := &attribute{
-		Name: name,
+		return nil, errors.Wrapf(err, "failed to get attribute sizes: %s...", names[0])
 	}
 
-	if attrSize > 0 {
-		buf := C.malloc(attrSize)
-		if buf == nil {
-			return nil, errors.New("failed to malloc buf")
+	// Now, create a slice of buffers to hold the values.
+	attrValues := make([]unsafe.Pointer, numAttr)
+	for i, size := range attrSizes {
+		if size < 1 {
+			return nil, errors.Errorf("failed to get attribute %s: size is %d", names[i], size)
 		}
```

Comment:
Memory leak - What if we allocate a buffer for the first attr in the list, but a later one fails here?

Reply:
Ah, good catch. I guess I need to move the deferred function declaration before the loop -- that will guarantee that it runs even if we exit here.

Reply:
Fixed.